File: algorithm/cpp/firstround/20.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

class Solution
{
public:
    bool isValid(string& s);
};

bool Solution::isValid(string& s)
{
    if(s.empty()) return true;
    int n = s.size();
    if(n&1) return false; //奇数个括号，肯定不匹配

    stack<char> stk;
    for (int i = 0; i < n; i++)
    {
        if(s[i]=='(' || s[i]=='[' || s[i]=='{') stk.push(s[i]);
        else
        {
            if(stk.empty()) return false;
            char tmp = stk.top();

            if((s[i]==')' && tmp=='(') || (s[i]==']' && tmp=='[') || (s[i]=='}' && tmp=='{'))
            {
                stk.pop();
            }
            else return false;
        }
    }
    return stk.empty();
}

int main()
{
    return 0;
}

File: leetcode_top100/go/firstround/20.go
package firstround

// 栈.时间复杂度O(n),空间复杂度O(n)
func isValid(s string) bool {
	stack := make([]int32, len(s))
	top := -1

	for _, ch := range s {
		if ch == '(' || ch == '[' || ch == '{' {
			top++
			stack[top] = ch
			continue
		}
		if (ch == ')' && stack[top] != '(') || (ch == ']' && stack[top] != '[') || (ch == '}' && stack[top] != '{') {
			return false
		}
		top--
	}
	return true
}
